### ISSUE TITLE
CD Audio stuttering under high load (esp w/ Voodoo emulation)

### ISSUE THREAD

**Issue by weirddan455** [dosbox-staging/dosbox-staging#3110]
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### What Operating System are you using?

None

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

Discussion started in #3103

@Kappa971 wrote:

> Let me explain in a simpler way what happens:
> If the 3dfx emulation is at full speed (60 fps) the CD Audio (music) runs at normal speed. If the emulation drops below 60 fps, the music slows down. That's all.
> I'm using audio files in OGG format, I haven't tried with WAV files. It could be that this behavior is unavoidable, I don't know, but I've reported it here in case something can be done about it.
> 
> With my CPU I can set around 140-150k cycles with the 3dfx emulation to play at full speed with decent performance, but I noticed that in the Carmageddon menu the emulation drops below 100% (50-55 fps) and the music slows down. The same thing sometimes happens with the pause menu of Screamer Rally. In practice it seems that the game menus are more challenging to emulate than the actual game 😆

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

**Comment by weirddan455** [comment]
@Grounded0 wrote:

> We recently introduced a fix to make sure Carmageddon doesn't stop playing its CD audio soundtrack after 20 minutes of playing. Probably good idea to look at it. This can be merged.

Can you please link to this fix?

**Comment by Grounded0** [comment]
Connected issues and PRs:

- #2622

**Comment by Grounded0** [comment]
@Kappa971 

Provide audio recording as a repro so we can confirm easily without going for a goose chase.

**Comment by Kappa971** [comment]
It's not just a Carmageddon problem, it seems that Audio CD playback is linked to the emulation speed, if this drops below 100% (so below 60 fps in fps counter), the music slows down. I've noticed that this usually happens to me in game menus (my CPU can easily handle 140k-150k cycles in game) and so in that case the music slows down.
I think you can have the same result with a low-performance CPU and too high cycles (even in game).
I was wondering if it is possible to separate the main emulation from the music playback.

**Comment by Grounded0** [comment]
Its easy to record audio for repro. Can you do that se we can hear the slowdown and upload or link it here.

**Comment by Grounded0** [comment]
You can see how to do a quality audio repro here:

- #3097

**Comment by weirddan455** [comment]
I can't reproduce a problem here.  I found an archive.org dump of the game here:  https://archive.org/download/carmageddon-redump/Carmageddon-PC-REDUMP/Carmageddon%20%28USA%29.zip

Ran it and the music sounds correct and the same speed whether in the menu or not.  I'll note that it's not .ogg files for the audio but .bin/.cue dumps of the original CD.

But I need more info:  Does it only happen with your ogg files?  If so, can you share them?  What settings are you using?  Does it only happen with Voodoo emulation?  What other games does this affect?  At the very least, some basic reproduction instructions.

**Comment by Kappa971** [comment]
It doesn't always happen:
51 FPS

https://github.com/dosbox-staging/dosbox-staging/assets/62349018/30172dd2-321f-4d37-95fd-84162c2abdb6

[CarmageddonMusicLags.zip](https://github.com/dosbox-staging/dosbox-staging/files/13329325/CarmageddonMusicLags.zip)

Screamer Rally in game (to demonstrate that it works well in game):
[ScreamerRallyInGame.zip](https://github.com/dosbox-staging/dosbox-staging/files/13329332/ScreamerRallyInGame.zip)

**Comment by weirddan455** [comment]
Thank you @Kappa971

That's not playing slowly, that's stuttering (which makes a lot more sense).  I still can't do anything about it without a consistent repo but this is something that offloading CD audio decoding/playing might help with which is something I plan to look at after the 0.81 release.

**Comment by johnnovak** [comment]
Please raise proper tickets @Kappa971 if you something to be looked at. No repro ticket, no fix. We created the issue template for this reason so we won't waste time doing this questioning all the time...

**Comment by Kappa971** [comment]
> Thank you @Kappa971
> 
> That's not playing slowly, that's stuttering (which makes a lot more sense). I still can't do anything about it without a consistent repo but this is something that offloading CD audio decoding/playing might help with which is something I plan to look at after the 0.81 release.

I'm not a programmer, but I think this is the problem.

Voodoo emulation slows down (cycles too high or I don't know what), SFX slows down/stutters (and I think this is normal) but it also slows down/stutters the CD Audio (OGG music).
On real hardware, if a game runs at 10 fps, does CD Audio slow down?

**Comment by johnnovak** [comment]
> On real hardware, if a game runs at 10 fps, does CD Audio slow down? I ask because I don't know.

CD Audio never "slows down" on hardware, but what you mean is stuttering due to dropouts as we've established. It never does that either (unless the CD is badly scratched 😅 ). With CD audio playback the CD-ROM drive literally acts as a CD player; the program sends it a single play track command, so it plays the track. In the DOS days the audio output of the CD drive was directly connected to the sound card which mixed it together with the rest of the sound card's output in hardware.

It was a full hardware, analog audio pathway, the audio never went through the software stack, so it was impossible for the software running on the CPU to affect the CD audio in any way.

**Comment by Kappa971** [comment]
>What settings are you using? 

Nothing special:
```
[sdl]
fullscreen = true

[dosbox]
machine  = svga_s3
memsize  = 64
vmemsize = auto
startup_verbosity = quiet

[cpu]
core      = dynamic
cputype   = auto
cycles    = 140000
```

> Does it only happen with your ogg files? If so, can you share them?

Ogg files were converted from a BIN CUE image: https://archive.org/details/carmageddon-italy.-7z_202302

> Does it only happen with Voodoo emulation?

No but with voodoo emulation it is much easier to reproduce the problem (with Screamer Rally in 65k high resolution I had to set 1,200,000 cycles).

> What other games does this affect?

I think all games with CD Audio.

**Comment by weirddan455** [comment]
Yeah, exactly as @johnnovak said on real hardware.

In Dosbox, the different emulation modules (CPU, OPL, Voodoo, etc.) run in serial at varying intervals.  Audio, I believe, runs every 1ms.  When you experience stuttering, it's likely because some other component (probably Voodoo in this case) is taking too long to finish its emulation "tick".

Because most things are being run from the main thread (including audio and I believe Voodoo as well?), this means the audio callback misses its normal 1ms interval so nothing gets played for a short while.  This leads to the very distinctive "stuttering" sound that you recorded.

Ideally, the entire audio system would be on a different thread.  If we miss a video frame or two, the player might not even notice.  If we miss audio frames, that stutter is very obvious.  At the very least, moving the decoding of .ogg files and such to its own thread should help as that's more CPU intensive than copying raw audio data from a .bin/.cue.

**Comment by kcgen** [comment]
Yup, that's right @weirddan455.

Every 1ms tick, the application or game can make calls to the emulated hardware IO ports/addresses or interrupts (as IO reads or writes). Those writes change the state of the device. We also query SDL for device inputs and those BIOS states are changes for the program to 'see' on the next 1ms tick.

For the audio, the mixer typically(*) requests 1ms worth of audio every tick from each audio device. More specifically, the amount requested is  based on how much a deficit the final mixed buffer is in.

Behind the scenes, SDL's has a thread that plays down the buffer we hand to it. 

`[mixer] prebuffer` controls how big that buffer is, in millseconds. By default it's 40 ms.

So in @Kappa971 case - the repeated heavy 3dfx calls and over-subscrubbed CPU cycle count is sending the buffer into > 40 ms deficits, repeatedly.

The mixer will try to recover by asking for all 40ms in the next tick to refill itself.. but that's even not enough when the system is busy for 60ms or more! Ouch.

**Comment by johnnovak** [comment]
Yeah, and running the CD Audio emulation in a separate thread would be a 100% solution, I think. That would faithfully mimic the real-world situation–CD Audio is a separate parallel process from the rest of the computer. We're already rendering MT-32 and FluidSynth audio in separate threads.

**Comment by weirddan455** [comment]
>  Threading the CD player will be a big help - but I fear we'll need threading at the mixer level as well to get those audio frames out to SDL even if the entire main loop is out to lunch. (Because currently the mixer itself is part of that 1ms loop - and the mixer is responsible for "fetching" audio frames from each device, include the CD player).

That's what I was alluding to in my comment.  Honestly, I wonder if we can just run the mixer inside SDL's callback function (which is already being run from a dedicated thread created by SDL).  I'd have to look at the code in more detail but that seems like the simplest solution and mixing audio is usually very fast to do in terms of CPU cycles.

>  Yeah, and running the CD Audio emulation in a separate thread would be a 100% solution, I think. That would faithfully mimic the real-world situation–CD Audio is a separate parallel process from the rest of the computer. We're already rendering MT-32 and FluidSynth audio in separate threads.

Agreed.  Parallelizing CD audio just makes sense.  The problem is you still have to sync up to the mixer which is running on the main thread.

**Comment by kcgen** [comment]
> I wonder if we can just run the mixer inside SDL's callback function (which is already being run from a dedicated thread created by SDL). I'd have to look at the code in more detail but that seems like the simplest solution and mixing audio is usually very fast to do in terms of CPU cycles.

Very interesting, @weirddan455!

**Comment by Grounded0** [comment]
One way to reduce stuttering would be just use plain .CUE/.BIN RedBook CD audio because its just a copy operation without decoding compression.

**Comment by johnnovak** [comment]
> That's what I was alluding to in my comment. Honestly, I wonder if we can just run the mixer inside SDL's callback function (which is already being run from a dedicated thread created by SDL). I'd have to look at the code in more detail but that seems like the simplest solution and mixing audio is usually very fast to do in terms of CPU cycles.

Neat idea and worth a try. Personally, I'd like to decouple SDL as much from the rest of the codebase as much as possible, so I'd prefer `std::thread` (similarly, I don't wanna let `SDL_Point` and `SDL_Rect` structs leak into our emulation code; I'm introducing a `Rectangle` struct for that in my upcoming PR).

(The elephant in the room is SDL's input handling, of course, and using their keycode consts everywhere...)

**Comment by weirddan455** [comment]
>  Personally, I'd like to decouple SDL as much from the rest of the codebase as much as possible

Usually a good idea to de-couple from libraries.  We can always write our own abstractions around any SDL'isms.  I was thinking it's more a matter of making sure we mix the audio when we actually need it.

We're already doing a bunch of fiddling with mixer variables inside SDL's callback and our current mixer solution seems a bit over-complicated in general.  I think the first place I would start in cleaning that up is #2829.  Use floating point everywhere and that trims down on a lot of the type conversion code I'm seeing in there.

**Comment by johnnovak** [comment]
> > Personally, I'd like to decouple SDL as much from the rest of the codebase as much as possible
> 
> Usually a good idea to de-couple from libraries. We can always write our own abstractions around any SDL'isms. I was thinking it's more a matter of making sure we mix the audio when we actually need it.
> 
> We're already doing a bunch of fiddling with mixer variables inside SDL's callback and our current mixer solution seems a bit over-complicated in general. I think the first place I would start in cleaning that up is #2829. Use floating point everywhere and that trims down on a lot of the type conversion code I'm seeing in there.

There's a lot of legacy cruft in `mixer.cpp`, agreed. Some remnants of the integer-based sample stream handling are still there, but I weeded out some parts of it already. By using floats consistently all those AddSamples templates should go away.

**Comment by Kappa971** [comment]
I would rename the title to something like "CD Audio stuttering if emulation slows down" or your choice. This isn't just a Carmageddon issue.

**Comment by Grounded0** [comment]
Could be closed by DOSBox Staging 0.82.0 RC1.

**Comment by johnnovak** [comment]
Yeah let's close it now as the fix is in `main` already.
